File: shoring/blackjackEngine/src/main/java/com/panopset/blackjackEngine/Player.java
package com.panopset.blackjackEngine;

import java.io.StringWriter;
import java.util.ArrayList;
import java.util.List;

public class Player {

  @Override
  public String toString() {
    StringWriter sw = new StringWriter();
    int i = 0;
    for (HandPlayer h : hands) {
      sw.append(String.format("Player #%d: cards:%s value: %d ", i, h.cards, h.getValue()));
      i++;
    }
    return sw.toString();
  }

  public List<HandPlayer> getHands() {
    return hands;
  }

  private final List<HandPlayer> hands = new ArrayList<>();

  public Player(final BlackjackGameEngine bge) {
    hands.add(new HandPlayer(bge));
  }

  public Player(Player player) {
    for (HandPlayer hand : player.getHands()) {
      hands.add(new HandPlayer(hand));
    }
  }

  public synchronized HandPlayer getActiveHand() {
    for (HandPlayer h : hands) {
      if (!h.isFinal()) {
        return h;
      }
    }
    return null;
  }

  public synchronized boolean isFinal() {
    return getActiveHand() == null;
  }

}

File: code/shoring/blackjackEngine/src/main/kotlin/com/panopset/blackjackEngine/Player.kt
package com.panopset.blackjackEngine

import com.panopset.compat.Logz
import java.io.StringWriter

class Player(val wager: Wager) {
    constructor(player: Player) : this( Wager(player.wager)) {
        hands.clear() // TODO: This is kinda lame.
        for (handPlayer in player.hands) {
            hands.add(HandPlayer( handPlayer))
        }
    }

    val hands = ArrayList<HandPlayer>()
    var isSettled = false
    init {
        if (hands.isEmpty()) {
            hands.add(HandPlayer( wager))
        }
    }

    override fun toString(): String {
        val sw = StringWriter()
        for ((i, h) in hands.withIndex()) {
            sw.append(String.format("Player #%d: cards:%s value: %d ", i, h.getBlackjackCards(), h.getHandValue()))
        }
        return sw.toString()
    }

    @get:Synchronized
    val activeHand: HandPlayer?
        get() {
            for (h in hands) {
                if (!h.isFinal()) {
                    return h
                }
            }
            return null
        }

    @get:Synchronized
    val isFinal: Boolean
        get() = activeHand == null
}
